tests/drivers/spi/spi_spis/src/main.c

```diff
@@ -0,0 +1,453 @@
+
+static struct k_poll_signal async_sig = K_POLL_SIGNAL_INITIALIZER(async_sig);
+static struct k_poll_event async_evt =
+	K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL, K_POLL_MODE_NOTIFY_ONLY, &async_sig);
+
+#define MEMORY_SECTION(node)                                                                       \
+	COND_CODE_1(DT_NODE_HAS_PROP(node, memory_regions),                                        \
+		    (__attribute__((__section__(                                                   \
+			    LINKER_DT_NODE_REGION_NAME(DT_PHANDLE(node, memory_regions)))))),      \
+		    ())
+
+static uint8_t spim_buffer[32] MEMORY_SECTION(DT_NODELABEL(dut_spi));
+static uint8_t spis_buffer[32] MEMORY_SECTION(DT_NODELABEL(dut_spis));
+
+struct test_data {
+	struct k_work_delayable test_work;
+	struct k_sem sem;
+	int spim_alloc_idx;
+	int spis_alloc_idx;
+	struct spi_buf_set sets[4];
+	struct spi_buf_set *mtx_set;
+	struct spi_buf_set *mrx_set;
+	struct spi_buf_set *stx_set;
+	struct spi_buf_set *srx_set;
+	struct spi_buf bufs[8];
+};
+
+static struct test_data tdata;
+
+/* Allocate buffer from spim or spis space. */
+static uint8_t *buf_alloc(size_t len, bool spim)
+{
+	int *idx = spim ? &tdata.spim_alloc_idx : &tdata.spis_alloc_idx;
+	uint8_t *buf = spim ? spim_buffer : spis_buffer;
+	size_t total = spim ? sizeof(spim_buffer) : sizeof(spis_buffer);
+
+	if (*idx + len > total) {
+		zassert_false(true);
+
+		return NULL;
+	}
+
+	uint8_t *rv = &buf[*idx];
+
+	*idx += len;
+
+	return rv;
+}
+
+static void work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct test_data *td = CONTAINER_OF(dwork, struct test_data, test_work);
+	int rv;
+
+	rv = spi_transceive_dt(&spim, td->mtx_set, td->mrx_set);
+
+	if (rv == 0) {
+		k_sem_give(&td->sem);
+	}
+}
+
+/** Copies data from buffers in the set to a single buffer which makes it easier
+ * to compare transmitted and received data.
+ *
+ * @param buf Output buffer.
+ * @param len Buffer length.
+ * @param set Set of buffers.
+ *
+ * @return Number of bytes copied.
+ */
+static int cpy_data(uint8_t *buf, size_t len, struct spi_buf_set *set)
+{
+	int idx = 0;
+
+	for (size_t i = 0; i < set->count; i++) {
+		size_t l = set->buffers[i].len;
+
+		if (len - idx >= l) {
+			memcpy(&buf[idx], set->buffers[i].buf, l);
+			idx += l;
+		} else {
+			return -1;
+		}
+	}
+
+	return idx;
+}
+
+/** Compare two sets.
+ *
+ * @param tx_set TX set.
+ * @param rx_set RX set.
+ * @param same_size True if it is expected to have the same amount of data in both sets.
+ *
+ * @return 0 if data is the same and other value indicate that check failed.
+ */
+static int check_buffers(struct spi_buf_set *tx_set, struct spi_buf_set *rx_set, bool same_size)
+{
+	static uint8_t tx_data[256];
+	static uint8_t rx_data[256];
+
+	if (!tx_set || !rx_set) {
+		return 0;
+	}
+
+	int rx_len = cpy_data(rx_data, sizeof(rx_data), rx_set);
```

Comment:
this habit of declaring variable in the middle of a code block is weird, even more when it's not even consistently applied.

Either all variables are declared at a code block start, or they are only declared when needed. But not both style mixed up.

It's a preference issue  so fully subjective, though the earlier helps a bit to evaluate stack usage of the code block.

Anyway, whatever you choose, try to be consistent.

(for instance these 2 variables in this function, vs the same ones in the function below. etc...)